daemons/ipa-kdb/ipa_kdb_principals.c

```diff
@@ -338,25 +337,84 @@ static void ipadb_validate_otp(struct ipadb_context *ipactx,
     if (dn == NULL)
         return;
     count = asprintf(&filter, ftmpl, dn, datetime, datetime);
-    ldap_memfree(dn);
-    if (count < 0)
+    if (count < 0) {
+        ldap_memfree(dn);
         return;
+    }
 
     /* Fetch the active token list. */
     kerr = ipadb_simple_search(ipactx, ipactx->base, LDAP_SCOPE_SUBTREE,
                                filter, (char**) attrs, &res);
     free(filter);
+    filter = NULL;
     if (kerr != 0 || res == NULL)
         return;
 
     /* Count the number of active tokens. */
     count = ldap_count_entries(ipactx->lcontext, res);
     ldap_msgfree(res);
 
-    /* If the user is configured for OTP, but has no active tokens, remove
-     * OTP from the list since the user obviously can't log in this way. */
-    if (count == 0)
-        *ua &= ~IPADB_USER_AUTH_OTP;
+    /*
+     * If there are no valid tokens then we need to remove the OTP flag,
+     * unless OTP is the only auth type allowed...
+     */
+    if (count == 0) {
+        if (*ua == IPADB_USER_AUTH_OTP) {
+            static const char *expired_ftmpl = "(&"
+                "(objectClass=ipaToken)(ipatokenOwner=%s)"
+                "(|(ipatokenNotAfter<=%s)(!(ipatokenNotAfter=*)))"
+            ")";
+
+            /* Remove the OTP flag for now */
+            *ua &= ~IPADB_USER_AUTH_OTP;
+
+            /*
+             * Ok, we "only" allow OTP, so if there is an expired token then
+             * add back the OTP flag as the server will double check the
+             * validity and reject the entire bind. Otherwise, this is the
+             * first time the user is authenticating and the user should be
+             * allowed to bind using its password
+             */
+            count = asprintf(&filter, expired_ftmpl, dn, datetime);
+            if (count < 0) {
```

Comment:
Re-using count is a nice optimization but it is confusing in context because it isn't counting anything. I think you can merge this into a single conditional if (asprint() < 0) { ... }

Reply:
> Re-using count is a nice optimization but it is confusing in context because it isn't counting anything. I think you can merge this into a single conditional if (asprint() < 0) { ... }

Agreed, the value serves no purpose